src/python_testing/mdns_discovery/mdns_discovery.py

```diff
@@ -0,0 +1,318 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+
+import asyncio
+
+from mdns_discovery.exceptions import DiscoveryNotPerformedError
+from zeroconf import ServiceStateChange, Zeroconf
+from zeroconf.asyncio import AsyncServiceBrowser, AsyncServiceInfo
+
+from .mdns_type_enum import MdnsType
+
+
+class MdnsDiscovery:
+    def __init__(self, tc):
+        """
+        Initializes the MdnsDiscovery instance with necessary configurations.
+
+        Args:
+            tc: A test case object containing node and fabric information.
+
+        Attributes:
+            _zc (Zeroconf): An instance of Zeroconf to manage mDNS operations.
+            _tc: The test case object.
+            _service_types (list): A list of mDNS service types to discover.
+            _discovered_services (dict): A dictionary to store discovered services.
+            _discovery_performed (bool): Flag to indicate if discovery has been performed.
+            dut_operational_service_name (str): The operational service name for the DUT (Device Under Test).
+        """
+        self._zc: Zeroconf = Zeroconf()
+        self._tc = tc
+        self._service_types = [mdns_type.value for mdns_type in MdnsType]
+        self._discovered_services = {}
+        self._discovery_performed = False
+        self.dut_operational_service_name = self._get_dut_operational_service_name()
+
+    async def discover(self, discovery_duration: float = 3) -> None:
```

Comment:
please attach unit of measurement for anything that has a unit. In this case `discovery_duration_sec` is probably apropriate. Should the default be `3.0` to ensure it is a float (in case someone tries to do float operations on it)?